File: typical90/python/069.py
import sys

input = sys.stdin.readline

N, K = map(int, input().split())
MOD = 1_000_000_007


def binpower(a: int, b: int) -> int:
    """べき乗を高速に計算する

    if 奇数: a*x^2
    if 偶数: x^2
    """
    ans = 1
    while b != 0:
        is_odd = b % 2 == 1
        if is_odd:
            ans = ans * a % MOD
        a = a * a % MOD
        b //= 2
    return ans


"""
K * (K - 1) * (K - 2) * (K - 2) * ... * (K - 2)
"""
if K == 1:
    print(1 if N == 1 else 0)

elif N == 1:
    print(K % MOD)

elif N == 2:
    print(K * (K - 1) % MOD)

else:
    # print(K * (K - 1) % MOD * binpower(K - 2, N - 2) % MOD)
    print(K * (K - 1) % MOD * pow(K - 2, N - 2, MOD) % MOD)

File: typical90/c++/069.cpp
#include <iostream>

const int MOD = 1000000007;

int binpower(int a, int b) {
  int ans = 1;
  while (b != 0) {
    if (b % 2 == 1) {
      ans = (long long)(ans)*a % MOD;
    }
    a = (long long)(a)*a % MOD;
    b /= 2;
  }
  return ans;
}

int main() {
  long long N;
  int K;
  std::cin >> N >> K;

  if (K == 1) {
    std::cout << (N == 1 ? 1 : 0) << std::endl;
  } else if (N == 1) {
    std::cout << K % MOD << std::endl;
  } else if (N == 2) {
    std::cout << (long long)(K) * (K - 1) % MOD << std::endl;
  } else {
    std::cout << (long long)(K) * (K - 1) * binpower(K - 2, N - 2) % MOD
              << std::endl;
  }
  return 0;
}
